eclass/kernel-install.eclass

```diff
@@ -459,15 +603,36 @@ kernel-install_install_all() {
 	local dir_ver=${1}
 	local kernel_dir=${EROOT}/usr/src/linux-${dir_ver}
 	local relfile=${kernel_dir}/include/config/kernel.release
+	local image_path=$(dist-kernel_get_image_path)
 	local module_ver
 	module_ver=$(<"${relfile}") || die
 
+	if [[ ${KERNEL_IUSE_GENERIC_UKI} ]]; then
+		# If the install image was built with USE=generic-uki, we only have uki.efi
+		# extract the initrd and kernel image
+		local uki="${image_path%/*}/uki.efi"
+		if [[ -f ${kernel_dir}/${uki} && ! -f ${kernel_dir}/${image_path} ]]; then
+			$(tc-getOBJCOPY) -O binary -j.linux "${kernel_dir}/${uki}" \
+				"${kernel_dir}/${image_path}" || die "Failed to extract kernel image"
+		fi
+		local initrd="${image_path%/*}/initrd"
+		if [[ -f ${kernel_dir}/${uki} && ! -f ${kernel_dir}/${initrd} ]] &&
+			use generic-uki; then
+				$(tc-getOBJCOPY) -O binary -j.initrd "${kernel_dir}/${uki}" \
+					"${kernel_dir}/${initrd}" || die "Failed to extract initramfs"
+		fi
+		if [[ -f ${kernel_dir}/${uki} ]] && ! use generic-uki; then
+			# If our gpkg was built with USE=generic-uki, but we don't want to use
+			# it, remove it, otherwise installkernel will pick it up automatically.
+			rm "${kernel_dir}/${uki}" || die "Failed to remove UKI"
+		fi
+	fi
+
 	local success=
 	# not an actual loop but allows error handling with 'break'
 	while :; do
 		nonfatal mount-boot_check_status || break
 
-		local image_path=$(dist-kernel_get_image_path)
 		if use initramfs && has_version "<=sys-kernel/installkernel-gentoo-7"; then
```

Comment:
How do you force rebuilding initramfs when the "generic" initramfs was originally installed?

Reply:
I mean, will `pkg_config` just work or does it need some extra logic? I'd imagine that if you install with `USE=generic-uki`, then run `emerge --config`, you'd want new initramfs anyway.

Reply:
> I mean, will `pkg_config` just work or does it need some extra logic? I'd imagine that if you install with `USE=generic-uki`, then run `emerge --config`, you'd want new initramfs anyway.

With `installkernel-gentoo` the dracut plugin always overwrites any existing initrd, this is why we IDEPEND on `USE="-dracut"`, this is the same behaviour as the eclass has with `installkernel-gentoo<=7`.
With `installkernel-systemd` the dracut plugin always uses an existing initrd before it tries to generate a new one (and if we want to change this we should do this upstream since this is where the plugin comes from).

`pkg-config` will just install the same initrd/uki again if `USE=generic-uki` (i.e. change nothing, but it might still be useful if the installed kernel is somehow removed by something external).

I don't think we can support `USE=generic-uki`, for which we must to ensure that `installkernel` does not overwrite the initd/uki we install, and at the same time allow `dist-kernel_reinstall_initramfs` or `emerge --config` to undo all of that. To be honest, in situations where users would want to regenerate the initrd (e.g. to include `sys-fs/zfs-kmod`) they should just disable `USE=generic-uki`

Reply:
> I mean, will `pkg_config` just work or does it need some extra logic? I'd imagine that if you install with `USE=generic-uki`, then run `emerge --config`, you'd want new initramfs anyway.

Thinking about this a bit more, I must say I disagree. If emerge fails to install the kernel because e.g. the ESP is full. Then I want and expect that `emerge --config ...` retries the same installation, if this would suddenly generate a new custom initrd this would be confusing.

Reply:
Yeah, I suppose that makes sense.